File: cli/src/core/ProcessedSourceStore.ts
/**
 * ProcessedSourceStore — the topic KB high-water mark, stored as the set of
 * already-ingested source IDs (NOT a timestamp). Decouples "has this source
 * been processed" from "what is its logical time" so out-of-order sources are
 * never skipped. Path: `topics/processed.json`, written via the active
 * StorageProvider (dual-write), mirroring CompiledStore.
 */

import { createLogger } from "../Logger.js";
import type { FileWrite } from "../Types.js";
import type { StorageProvider } from "./StorageProvider.js";
import { resolveStorage } from "./SummaryStore.js";
import type { ProcessedSet, SourceRef, SourceType } from "./TopicKBTypes.js";

const log = createLogger("ProcessedSourceStore");
const PROCESSED_PATH = "topics/processed.json";

/** A fresh set with all four buckets present. */
export function emptyProcessedSet(): ProcessedSet {
	return { schemaVersion: 1, processed: { summary: [], plan: [], note: [], userfile: [] } };
}

/** Reads `topics/processed.json`; missing or unparseable → empty set (never throws). */
export async function readProcessedSet(cwd?: string, storage?: StorageProvider): Promise<ProcessedSet> {
	const resolved = resolveStorage(storage, cwd);
	const raw = await resolved.readFile(PROCESSED_PATH);
	if (!raw) return emptyProcessedSet();
	try {
		const parsed = JSON.parse(raw) as Partial<ProcessedSet>;
		const p: Partial<Record<SourceType, string[]>> = parsed.processed ?? {};
		return {
			schemaVersion: 1,
			processed: {
				summary: p.summary ?? [],
				plan: p.plan ?? [],
				note: p.note ?? [],
				userfile: p.userfile ?? [],
			},
		};
	} catch {
		log.warn("Failed to parse %s — treating as empty", PROCESSED_PATH);
		return emptyProcessedSet();
	}
}

/**
 * True when `ref` (by type+id) is already in the set.
 *
 * Membership is a linear `Array.includes` per bucket — fine at current scale
 * (bounded by source count). If buckets grow large enough that batch
 * `addProcessed` becomes a hotspot, switch the in-memory representation to a
 * per-bucket `Set` (the on-disk JSON stays an array for readability; see the
 * parent spec's deferred compaction note).
 */
export function hasProcessed(set: ProcessedSet, ref: SourceRef): boolean {
	return set.processed[ref.type].includes(ref.id);
}

/** Returns a new set with `refs` added (idempotent, does not mutate `set`). */
export function addProcessed(set: ProcessedSet, refs: ReadonlyArray<SourceRef>): ProcessedSet {
	const next: Record<SourceType, string[]> = {
		summary: [...set.processed.summary],
		plan: [...set.processed.plan],
		note: [...set.processed.note],
		userfile: [...set.processed.userfile],
	};
	for (const ref of refs) {
		if (!next[ref.type].includes(ref.id)) next[ref.type].push(ref.id);
	}
	return { schemaVersion: 1, processed: next };
}

/** Persists the set via the active StorageProvider. */
export async function saveProcessedSet(set: ProcessedSet, cwd?: string, storage?: StorageProvider): Promise<void> {
	const resolved = resolveStorage(storage, cwd);
	const files: FileWrite[] = [{ path: PROCESSED_PATH, content: JSON.stringify(set, null, "\t") }];
	await resolved.writeFiles(files, "Update topic KB processed-source set");
}

File: intellij/src/main/kotlin/ai/jolli/jollimemory/core/ProcessedSourceStore.kt
package ai.jolli.jollimemory.core

/**
 * ProcessedSourceStore — the topic-KB high-water mark, stored as the set of
 * already-ingested source IDs (NOT a timestamp). Decouples "has this source
 * been processed" from "what is its logical time" so out-of-order sources are
 * never skipped. Path: `topics/processed.json`, written via the [StorageProvider].
 *
 * Kotlin port of `cli/src/core/ProcessedSourceStore.ts`.
 */
object ProcessedSourceStore {

    private val log = JmLogger.create("ProcessedSourceStore")
    private const val PROCESSED_PATH = "topics/processed.json"

    /** A fresh set with all four buckets present, in canonical order. */
    fun emptyProcessedSet(): ProcessedSet =
        ProcessedSet(schemaVersion = 1, processed = canonicalBuckets(emptyMap()))

    /** Reads `topics/processed.json`; missing or unparseable → empty set (never throws). */
    fun readProcessedSet(storage: StorageProvider): ProcessedSet {
        val raw = storage.readFile(PROCESSED_PATH) ?: return emptyProcessedSet()
        val parsed = TopicJson.parse(raw, ProcessedSet::class.java)
        if (parsed == null) {
            log.warn("Failed to parse %s — treating as empty", PROCESSED_PATH)
            return emptyProcessedSet()
        }
        return ProcessedSet(schemaVersion = 1, processed = canonicalBuckets(parsed.processed))
    }

    /** True when [ref] (by type+id) is already in the set. */
    fun hasProcessed(set: ProcessedSet, ref: SourceRef): Boolean =
        set.processed[ref.type]?.contains(ref.id) == true

    /** Returns a new set with [refs] added (idempotent, does not mutate [set]). */
    fun addProcessed(set: ProcessedSet, refs: List<SourceRef>): ProcessedSet {
        val next = LinkedHashMap<String, MutableList<String>>()
        for (type in SourceType.ALL) next[type] = (set.processed[type] ?: emptyList()).toMutableList()
        for (ref in refs) {
            val bucket = next.getOrPut(ref.type) { mutableListOf() }
            if (!bucket.contains(ref.id)) bucket.add(ref.id)
        }
        return ProcessedSet(schemaVersion = 1, processed = next.mapValues { it.value.toList() })
    }

    /** Persists the set via the provider (byte-stable tab-indented JSON). */
    fun saveProcessedSet(set: ProcessedSet, storage: StorageProvider) {
        storage.writeFiles(
            listOf(FileWrite(path = PROCESSED_PATH, content = TopicJson.stringify(set))),
            "Update topic KB processed-source set",
        )
    }

    /** Rebuilds the bucket map in canonical `summary, plan, note, userfile` order. */
    private fun canonicalBuckets(src: Map<String, List<String>>): Map<String, List<String>> {
        val out = LinkedHashMap<String, List<String>>()
        for (type in SourceType.ALL) out[type] = src[type] ?: emptyList()
        return out
    }
}
